winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,198 @@
+#include <stdio.h>
+#include <windows.h>
+
+/**
+ * To determine the address of kernel32!CtrlRoutine, we need to use
+ * dbghelp.dll. But we want to avoid linking statically to that library because
+ * the normal operation of cygwin-console-helper.exe (i.e. to allocate a hidden
+ * Console) does not need it.
+ *
+ * Therefore, we declare the SYMBOL_INFOW structure here, load the dbghelp
+ * library via LoadLibraryExA() and obtain the SymInitialize(), SymFromAddrW()
+ * and SymCleanup() functions via GetProcAddr().
+ * 
+ * After that a remote thread is injected into that process
+ * to send the exit code.
+ */
+
+#include <dbghelp.h>
+
+static int pid;
+static uintptr_t exit_code;
+static BOOL do_print_address;
+
+/* Avoid fprintf(), as it would try to reference '__getreent' */
+static void
+output (BOOL error, const char *fmt, ...)
+{
+  va_list ap;
+  char buffer[1024];
+
+  va_start (ap, fmt);
+  vsnprintf (buffer, sizeof(buffer) - 1, fmt, ap);
+  buffer[sizeof(buffer) - 1] = '\0';
+  va_end (ap);
+  WriteFile (GetStdHandle(error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE),
+	      buffer, strlen (buffer), NULL, NULL);
+}
+
+static int
+inject_remote_thread_into_process(HANDLE process, LPTHREAD_START_ROUTINE address, uintptr_t exit_code)
+{
+  int res = -1;
+
+  if (!address)
+    return res;
+  DWORD thread_id;
+  HANDLE thread = CreateRemoteThread(process, NULL, 1024 * 1024, address,
+                                     (PVOID)exit_code, 0, &thread_id);
+  if (thread)
+  {
+    /*
+      * Wait up to 10 seconds (arbitrary constant) for the thread to finish;
+      * After that grace period, fall back to terminating non-gently.
+      */
+    if (WaitForSingleObject(thread, 10000) == WAIT_OBJECT_0)
+      res = 0;
+    CloseHandle(thread);
+  }
+
+  return res;
+}
+
+static WINAPI BOOL
+ctrl_handler(DWORD ctrl_type)
+{
+  unsigned short count;
+  void *address;
+  HANDLE process;
+  PSYMBOL_INFOW info;
+  DWORD64 displacement;
+
+  count = CaptureStackBackTrace (1l /* skip this function */,
+			         1l /* return only one trace item */,
+				 &address, NULL);
+  if (count != 1)
+    {
+      output (1, "Could not capture backtrace\n");
+      return FALSE;
+    }
+
+  process = GetCurrentProcess ();
+  if (!SymInitialize (process, NULL, TRUE))
+    {
+      output (1, "Could not initialize symbols\n");
+      return FALSE;
+    }
+
+  info = (PSYMBOL_INFOW)
+    malloc (sizeof(*info) + MAX_SYM_NAME * sizeof(wchar_t));
+  if (!info)
+    {
+      output (1, "Could not allocate symbol info structure\n");
+      return FALSE;
+    }
+  info->SizeOfStruct = sizeof(*info);
+  info->MaxNameLen = MAX_SYM_NAME;
+
+  if (!SymFromAddrW (process, (DWORD64)(intptr_t)address, &displacement, info))
+    {
+      output (1, "Could not get symbol info\n");
+      SymCleanup(process);
+      return FALSE;
+    }
+
+  if (do_print_address)
+  {
```

Comment:
When adding code, it is always good to imitate the coding style of the surrounding code. It's like the saying "When in Rome, do as the Romans do".

In this instance, the curly brace should be indented two more spaces than the preceding `if` statement.

(For the record, I dislike the GNU coding convention, but my taste does not matter at all when I contribute to Cygwin, it matters more that the coding style is consistent because it would otherwise be disruptive to the reader, which includes future me.)

Reply:
Fine. I was doing this like "code first, make it work, style last", so I would format them accordingly after everything is working.

Reply:
Style fixes are _the worst_ when you have to perform an interactive rebase. They cause not only merge conflicts when moving the style fix to the correct commit, no, _another_ merge conflict (at _least_ one, that is) will happen when the incorrect style is expected in later commits' diff context.

If you plan on cleaning this up, the sooner you do, the less hassle it will be.

Reply:
Oh right, I will clean them up soon.

Reply:
Is there a style guide somewhere for this project? I am, slightly unfamiliar with that.
( Also, this is a side-effect of vscode automatically formatting things, if it can be automatically formatted it would be better)

Reply:
Found there was an option in vscode to add GNU format, and it sort of works good.

Reply:
OK fixed it with a force push.